geemap/toolbar.py

```diff
@@ -27,211 +31,130 @@
-                button_style="primary",
-                icon=icons[i],
-                tooltip=tooltips[i],
-            )
-            for i in range(len(all_tools))
-        ]
-        self.toggle_widget = self.all_widgets[len(main_tools)] if extra_tools else None
+    # The accessory widget.
+    accessory_widgets = map_widgets.TypedTuple(
+        trait=traitlets.Instance(widgets.Widget),
+        help="The accessory widget",
+    ).tag(sync=True, **widgets.widget_serialization)
 
-        # We start with a collapsed grid of just the main tools and the toggle one.
-        self.grid = widgets.GridBox(
-            children=self.all_widgets[: self.num_collapsed_tools],
-            layout=widgets.Layout(
-                width="109px",
-                grid_template_columns=(self.ICON_WIDTH + " ") * self.NUM_COLS,
-                grid_template_rows=(self.ICON_HEIGHT + " ") * self.num_rows_collapsed,
-                grid_gap="1px 1px",
-                padding="5px",
-            ),
-        )
+    # The list of main tools.
+    main_tools = map_widgets.TypedTuple(
+        trait=traitlets.Instance(widgets.Widget),
+        help="List of main tools",
+    ).tag(sync=True, **widgets.widget_serialization)
 
-        def curry_callback(callback, should_reset_after, widget, item):
-            def returned_callback(change):
-                if change["type"] != "change":
-                    return
-                callback(self.host_map, change["new"], item)
-                if should_reset_after:
-                    widget.value = False
+    # The list of extra tools.
+    extra_tools = map_widgets.TypedTuple(
+        trait=traitlets.Instance(widgets.Widget),
+        help="List of extra tools",
+    ).tag(sync=True, **widgets.widget_serialization)
 
-            return returned_callback
+    # Whether the toolbar is expanded.
+    expanded = traitlets.Bool(False).tag(sync=True)
 
-        for id, widget in enumerate(self.all_widgets):
-            all_tools[id].toggle_button = widget
-            widget.observe(
-                curry_callback(callbacks[id], resets[id], widget, all_tools[id]),
-                "value",
-            )
+    # The currently selected tab.
+    tab_index = traitlets.Int(0).tag(sync=True)
 
-        self.toolbar_button = widgets.ToggleButton(
-            value=False,
-            tooltip="Toolbar",
-            icon="wrench",
-            layout=widgets.Layout(
-                width="28px", height="28px", padding="0px 0px 0px 4px"
-            ),
-        )
+    _TOGGLE_EXPAND_ICON = "add"
+    _TOGGLE_EXPAND_TOOLTIP = "Expand toolbar"
+    _TOGGLE_COLLAPSE_ICON = "remove"
+    _TOGGLE_COLLAPSE_TOOLTIP = "Collapse toolbar"
 
-        self.layers_button = widgets.ToggleButton(
-            value=False,
-            tooltip="Layers",
-            icon="server",
-            layout=widgets.Layout(height="28px", width="72px"),
+    def __init__(self, host_map, main_tools, extra_tools, accessory_widgets):
```

Comment:
While you're here, could you add type annotations to the constructor?

Reply:
Yes! Done. Also changed up some types to make more sense.